os/drivers/serial/serial.c

```diff
@@ -125,6 +132,87 @@ static const struct file_operations g_serialops = {
  * Private Functions
  ************************************************************************************/
 
+#ifdef CONFIG_DYNAMIC_LOGS_STORING
+struct log_net_chunk_s *get_log_net_head(void) {
+	return log_net_head;
+}
+
+static int read_counter = 0;
+
+void read_log_net(void) {
+        struct log_net_chunk_s *next_chunk = log_net_head;
+	struct mm_heap_s *myheap = mm_get_heap(log_net_head); /* get heap details using the head of the chunks linked list */
+
+                if (!myheap) {
+                        return; /* how to proceed if myheap is NULL? */
+                }
+
+                size_t free_size = myheap->mm_heapsize - myheap->total_alloc_size; /* get the free size */
+        lldbg("Read output %dth time!! ---> starts here <--- log size : %d, free heap : %d\n", read_counter, log_net_size, free_size);
+
+	while (next_chunk != NULL) {
+                for (int i = 0; i < next_chunk->size; i++) { /* printing the data to verify storing of logs, later to be replaced by network code to send to server*/
+                        lldbg_noarg("%c", next_chunk->arr[i]);
+                }
+                struct log_net_chunk_s *tmp = next_chunk;
+                next_chunk = next_chunk->next;
+                free(tmp); /* once we read the stored data, free the node */
+        }
+	lldbg("Read output %dth time!! ---> ends here <---\n", read_counter);
+        log_net_head = NULL;
+}
+
+static void add_char_to_log_net(char ch)
+{
+	if (log_net_head == NULL) { /* linked list is empty */
+		/* no chunks are present so we allocate new chunk */
+		struct log_net_chunk_s *node = (struct log_net_chunk_s *)kmm_malloc(sizeof(struct log_net_chunk_s));
+		node->next = NULL;
+		node->size = 1;
+		node->arr[0] = ch;
+		log_net_head = node;
+		log_net_tail = node;
+		/* set the size of chunks to 1 chunk */
+		log_net_size = sizeof(struct log_net_chunk_s);
+
+	} else { /* need to check if the current chunks size if over max_log_size or greater than x% of free heap */
+
+		struct mm_heap_s *myheap = mm_get_heap(log_net_head); /* get heap details using the head of the chunks linked list */
+
+		if (!myheap) {
+			return; /* how to proceed if myheap is NULL? */
+		}
+
+		size_t free_size = myheap->mm_heapsize - myheap->total_alloc_size; /* get the free size */
+
+		while (log_net_size / free_size * 100 > CONFIG_MAX_FREE_HEAP_PERCENT || log_net_size > CONFIG_MAX_LOG_SIZE_KB * 1024) {
+
+			struct log_net_chunk_s *node = log_net_head;
+			log_net_head = node->next;
+			free(node);
```

Comment:
If you use the memory allocation api as `kmm_xxx`, then we should use the free api as `kmm_free`.
Please change it.